### PR TITLE
Refactor Selection framework: combine Slice, Slices, SlicesBuilder into one.

### PR DESCRIPTION
**Pull request by stellama0208**
- Only keep Slices.
 - Remove Slice: The main difference between Slice and Slices is the different SelectionView Composite returned (defining how the info box beneath looks like), handle this treatment in the new code by returning corresponding Composite based on slice count.
 - Remove SlicesBuilder: Decouple something out of the combining logic and make combining simple enough, so that we can see "combinable" a natural attribute of Slices itself. This removal will clean out some unnecessary transformations and make combination process more efficient.
 - Bug: b/154967330.

### CODE REVIEW

File: gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
 import com.google.gapid.perfetto.views.State;
 
 import java.util.Arrays;
+import org.eclipse.swt.SWT;
```
Comment (rschiu) [inline gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java:40]: Unused?

File: gapic/src/main/com/google/gapid/perfetto/views/CpuSlicesSelectionView.java [hunk 2]
```diff
@@ -35,16 +37,18 @@
  * Displays information about a list of selected CPU slices.
  */
 public class CpuSlicesSelectionView extends Composite {
```
Comment (silence-do-good) [inline gapic/src/main/com/google/gapid/perfetto/views/CpuSlicesSelectionView.java:39]: Now that we have only Slices class, do we want to rename either this as Multi Selection view or the SliceSelectionView as SingleSliceSelectionView or something similar?

Comment (silence-do-good) [reply gapic/src/main/com/google/gapid/perfetto/views/CpuSlicesSelectionView.java:39]: Alternatively, we can have only the SlicesSelectionView composite and based on the count, show either the single selection or the multi-selection.

File: gapic/src/main/com/google/gapid/perfetto/views/ThreadPanel.java [hunk 3]
```diff
@@ -412,11 +409,11 @@ public void computeSelection(CombiningBuilder builder, Area area, TimeSpan ts) {
     }
 
     if (startDepth == 0) {
-      builder.add(Selection.Kind.ThreadState,
-          transform(track.getStates(ts), ThreadTrack.StateSlicesBuilder::new));
-      builder.add(Selection.Kind.Cpu, transform(track.getCpuSlices(ts), r -> {
-        r.forEach(s -> state.addSelectedThread(state.getThreadInfo(s.utid)));
-        return new CpuTrack.SlicesBuilder(r);
+      // Extra transform here solving a java generic type recognition problem.
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/ThreadPanel.java:412]: I've sent you a PR (https://github.com/stellama0208/agi/pull/1) to get rid of this.

**Comment by stellama0208** [comment]
The above force push was only to resolve merge conflict with AGI tot.

File: gapic/src/main/com/google/gapid/perfetto/views/ThreadPanel.java (hunk 3 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/ThreadPanel.java]: Thanks a lot for helping solve this complicated generic thing! A lot to learn in this PR!
Somehow on my machine, there's still a little compilation problem after this patch, so in "Address comments #1" I added several more type conversions, hope they are tolerable!

File: gapic/src/main/com/google/gapid/perfetto/views/CpuSlicesSelectionView.java (hunk 2 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/CpuSlicesSelectionView.java:46]: Thanks for this great idea! I've done it following your second idea in "Address comments #2". File number is reduced thanks to your suggestion!

File: gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java [hunk 4]
```diff
@@ -42,31 +42,35 @@
   private static final int PROPERTIES_PER_PANEL = 8;
   private static final int PANEL_INDENT = 25;
 
-  public FrameEventsSelectionView(Composite parent, State state, FrameEventsTrack.Slice slice) {
+  public FrameEventsSelectionView(Composite parent, State state, FrameEventsTrack.Slices slice) {
     super(parent, SWT.NONE);
     setLayout(withMargin(new GridLayout(3, false), 0, 0));
 
+    if (slice.count != 1) {
+      throw new IllegalArgumentException("Slice count != 1. Should only use FrameEventsSelectionView for a single FrameEvents slice.");
```
[the remark this answers is not in the document]
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/FrameEventsSelectionView.java]: This one is implicitly solved with "Address comments #2", which does count checking and composite creation in the same place.

File: gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java [hunk 5]
```diff
@@ -225,77 +214,37 @@ public Data(DataRequest request, long[] ids, long[] starts, long[] ends, long[]
...
-      this.endState = endState;
-      this.priority = priority;
-    }
-
-    public Slice(QueryEngine.Row row) {
-      this(row.getLong(0), row.getLong(1), row.getLong(2),
-          row.getInt(3), row.getLong(4), row.getLong(5),
-          ThreadState.of(row.getString(6)), row.getInt(7));
+  public static class Slices implements Selection<Slices> {
+    public int count = 0;
```
[the remark this answers is not in the document]
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java]: Thanks for suggestion, done in "Address comments #3".

File: gapic/src/main/com/google/gapid/perfetto/models/SliceTrack.java [hunk 6]
```diff
@@ -319,18 +234,63 @@ public RenderStageInfo(long frameBufferHandle, String frameBufferName, long rend
...
+      this.argsets.add(argset);
+      this.sliceKeys.add(id);
+    }
+
+    public ThreadInfo getThreadAt(int index) {
+      return ThreadInfo.EMPTY;
+    }
+
+    public RenderStageInfo getRenderStageInfoAt(int index) {
+      return RenderStageInfo.EMPTY;
```
[the remark this answers is not in the document]
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/SliceTrack.java]: Done in "Address comments #4". 

This was written this way as two default methods, extended class `GpuSlices` has real implementation for `getRenderStageInfoAt` and `ThreadSlices` has real implementation for `getThreadAt`.
Change related logic a bit in the new commit.

File: gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java [hunk 7]
```diff
@@ -310,56 +259,53 @@ public boolean contains(Long key) {
...
 
     @Override
-    public SlicesBuilder combine(SlicesBuilder other) {
-      for (Slice s : other.slices) {
-        if (!this.sliceKeys.contains(s.id)) {
-          this.slices.add(s);
-          this.sliceKeys.add(s.id);
+    public Slices combine(Slices other) {
+      for (int i = 0; i < other.count; i++) {
+        if (!this.sliceKeys.contains(other.ids.get(i))) {
```
[the remark this answers is not in the document]
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java:277]: Actually this scenario happens:
1. CTRL+SHIFT+DRAG on existing selections.
2. CTRL+CLICK on one selection multiple times.
If this logic doesn't exist, in the info box, we would see duplicated rows corresponding to the same selection.

File: gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java (hunk 1 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/CpuTrack.java]: Thanks for noting, have removed unused imports in new commits.

**Comment by stellama0208** [comment]
The above comments were addressed in new commits, ready for a new look, thanks! :)

**Pull request merged**